### DISCUSSION TITLE
Reconsider how network device creation from tap file descriptors is exposed over the API

### DISCUSSION THREAD

**Discussion by alyssais** [cloud-hypervisor/cloud-hypervisor#2514] [Q&A] [answered]
Specifying file descriptors by number makes sense for the CLI, because the user can set up the file descriptors however they want before `exec`-ing cloud-hypervisor.  And adding network devices from tap FDs at runtime is certainly a useful feature, but the way it's currently implemented makes it almost useless.

Currently, the API works just like CLI. You send it `{"fds":[3]}`, for example, and it'll try to use descriptor 3 as a tap to back a new virtio-net device.  But that requires there to already be a tap on descriptor 3 in cloud-hypervisor's descriptor table -- it'd have to have been put there before cloud-hypervisor was started, so the only use case this serves is the one where you have a tap device available and set up ahead of time, but for some reason don't want to attach the virtio-net device to the guest until later.  The layout of a running process's file descriptors should really be an implementation detail.  Exposing this makes some abuses possible too: if I have a VM with a serial console, I can send it `{"fds":[1]}`, which will cause it to try to do a tuntap ioctl on stdout, and then close it when the ioctl fails.  Now stdout is closed, but cloud-hypervisor still thinks it's open.  When it does anything else that opens a file descriptor, that new descriptor will become descriptor 1, so further writes to the serial console will instead go to that new descriptor, causing who knows what to happen.

It would be much more useful, and more correct, to allow other processes to actually send file descriptors to cloud-hypervisor using SCM_RIGHTS. (I have a use-case that could really use this feature!)  Failing that, I think that the potential utility of exposing `fds` over the API is so small that it's far outweighed by the downsides of letting other processes tell cloud-hypervisor how to interpret file descriptors at runtime, that it should probably be CLI-only.

**Comment by sboeuf** [comment]
Hi @alyssais, thanks for opening this issue. Could you move this to [Github discussions](https://github.com/cloud-hypervisor/cloud-hypervisor/discussions) ?

So you're suggesting to use ancillary data and the OOB mechanism to send some file descriptors to Cloud Hypervisor, right? The problem I see with this approach is that you need some kind of protocol to be defined by Cloud Hypervisor in order to know, as a separate process, how to connect and send these file descriptors.

Any management layer process making use of a VMM like Cloud Hypervisor is in charge of spawning the VMM's process, therefore it is fully capable of opening the TAP itself and share it with the VMM. I'm not sure why you can't leverage this in your case?

**Comment by alyssais** [comment]
> Hi @alyssais, thanks for opening this issue. Could you move this to [Github discussions](https://github.com/cloud-hypervisor/cloud-hypervisor/discussions) ?

I don't see an option in GitHub to do that. Perhaps it requires repository write access?

> So you're suggesting to use ancillary data and the OOB mechanism to send some file descriptors to Cloud Hypervisor, right? The problem I see with this approach is that you need some kind of protocol to be defined by Cloud Hypervisor in order to know, as a separate process, how to connect and send these file descriptors.

Correct.  I'm not sure I understand the second sentence here.  Can you rephrase the problem?

> Any management layer process making use of a VMM like Cloud Hypervisor is in charge of spawning the VMM's process, therefore it is fully capable of opening the TAP itself and share it with the VMM. I'm not sure why you can't leverage this in your case?

What do you mean by "share it with the VMM"?  Do you mean just tell cloud-hypervisor the name of the TAP and have it open it itself?  Or do you mean open a TAP file descriptor and share that with the VM?  In the former case, it means I can't take advantage of the benefits of using a file descriptor for the TAP device.  One such benefit is that I can set the TAP to be non-persistent before giving it to cloud-hypervisor, and then the kernel will clean it up when cloud-hypervisor closes it.  In the latter case, I don't believe it's possible to add file descriptors to another process at runtime without cooperation from that process, and even if it were, the sending process wouldn't know the descriptor number, so still wouldn't be able to tell cloud hypervisor which descriptor to look at.

**Comment by sboeuf** [comment]
>I don't see an option in GitHub to do that. Perhaps it requires repository write access?

Can you see/click this [link](https://github.com/cloud-hypervisor/cloud-hypervisor/discussions/new) ?

>Correct. I'm not sure I understand the second sentence here. Can you rephrase the problem?

Nevermind, I think I was thinking in terms of CLI while you were explaining the problem with the HTTP API. So basically yes I think we could leverage OOB mechanism to send some fds over the API socket. Cloud Hypervisor could expect to receive such ancillary data only in case `net` would be present.
We could also decide to have a dedicated option in order to send fds through the HTTP API, and we could identify these fds with unique identifiers. Then from the `net` option, we would refer to this id so that Cloud Hypervisor can retrieve the right fd.

**Comment by alyssais** [comment]
>>I don't see an option in GitHub to do that. Perhaps it requires repository write access?
>
> Can you see/click this [link](https://github.com/cloud-hypervisor/cloud-hypervisor/discussions/new) ?

Ah, yes, I can create a new discussion, but I can't convert this issue
into a discussion with this feature:
https://github.com/github/feedback/discussions/2952

Did you mean for me to just close this issue and open a new discussion?

>>Correct. I'm not sure I understand the second sentence here. Can you rephrase the problem?
>
> Nevermind, I think I was thinking in terms of CLI while you were
> explaining the problem with the HTTP API. So basically yes I think we
> could leverage OOB mechanism to send some fds over the API
> socket. Cloud Hypervisor could expect to receive such ancillary data
> only in case `net` would be present.  We could also decide to have a
> dedicated option in order to send fds through the HTTP API, and we
> could identify these fds with unique identifiers. Then from the `net`
> option, we would refer to this id so that Cloud Hypervisor can
> retrieve the right fd.

By "OOB", do you mean MSG_OOB?  If so, unix(7) says:

> UNIX domain sockets do not support the transmission of out-of-band
> data (the MSG_OOB flag for send(2) and recv(2)).

But that seems to be a different thing to sending file descriptors with
SCM_RIGHTS, which I'd expect an API client to just add to its sendmsg
call when it sends the HTTP request to the socket.

**Comment by sboeuf** [comment]
@alyssais 

I migrated the issue into a GH discussion, I didn't realize it required certain permissions to do so.

About OOB, I actually meant ancillary data/cmsg/scm rights, so we're talking about the same thing. I think your point is totally valid and the way we accept fd number at the moment is only valid for a caller opening the fd beforehand and passing them to CH at the time CH is spawned.

The way to implement this feature so that it could be beneficial for any future use case (not specific to networking) would be to introduce a dedicated API to let the caller send a file descriptor. This must return the file descriptor number from CH perspective, as this would be the way for the caller to refer to this file descriptor at a later time (when adding a new network for instance).
I'm just not sure if we can reuse the `api-socket` for this purpose, as it might introduce some complexity along with the HTTP protocol handling. It might be easier to define a dedicated UNIX socket for this, that might be defined at the time the VM is started or maybe through the API call directly.
What do you think?

Also, you mentioned about having a use case for this, can you maybe share some details?

**Comment by alyssais** [comment]
> I migrated the issue into a GH discussion, I didn't realize it required certain permissions to do so.
>
> About OOB, I actually meant ancillary data/cmsg/scm rights, so we're talking about the same thing. I think your point is totally valid and the way we accept fd number at the moment is only valid for a caller opening the fd beforehand and passing them to CH at the time CH is spawned.
>
> The way to implement this feature so that it could be beneficial for any future use case (not specific to networking) would be to introduce a dedicated API to let the caller send a file descriptor. This must return the file descriptor number from CH perspective, as this would be the way for the caller to refer to this file descriptor at a later time (when adding a new network for instance).
> I'm just not sure if we can reuse the `api-socket` for this purpose, as it might introduce some complexity along with the HTTP protocol handling. It might be easier to define a dedicated UNIX socket for this, that might be defined at the time the VM is started or maybe through the API call directly.
> What do you think?

From a theoretical point of view, it seems a shame to have two sockets
when one would work (and it definitely would -- I've previously written
a Serde extension that integrates file descriptor serialization into
Unix sockets using the normal serde API for an experiment I was doing
with crosvm:
https://spectrum-os.org/git/crosvm/tree/msg_socket2?h=interguest).

But in practice, it's probably makes sense to have two seperate sockets,
because approximately nothing that speaks HTTP will be able to send file
descriptors, and even HTTP libraries probably usually handle the socket
communication themselves.  So what you propose, with a seperate socket
that can be sent file descriptors and respond with identifiers that can
be passed to the HTTP API makes sense.

> Also, you mentioned about having a use case for this, can you maybe share some details?

Sure.  I have a long-lived VM, and I want to be able to at a later point
start a short-lived VM, and connect it to the long-lived one using a
bridge on the host.  To do this, I need to add a TAP to the long-lived
VM while it's running.  Once the short-lived VM is done, I need to clean
up, so I need to be able to remove the TAP from the VM and then delete
it.

As it stands with cloud-hypervisor 0.14.1, if I delete the TAP
immediately after getting a response to my vm.remove-device request,
I'll get a read error from cloud-hypervisor.  I haven't checked, but I
assume this is because cloud-hypervisor isn't actually done with the TAP
when it responds to the request.

If I could send cloud-hypervisor a TAP file descriptor, I could create a
non-persistent TAP, send it to cloud-hypervisor, and then avoid having
to delete the TAP entirely, because the kernel would clean it up when
cloud-hypervisor closed it.

It could be argued that what I saw with the read errors is a bug, but
even so being able to send a file descriptor instead would be cleaner,
because then I wouldn't have to worry about deleting the TAP at all.

**Comment by sboeuf** [comment, marked as answer]
@alyssais after some thinking and internal discussions, here is the design decision we took: https://github.com/cloud-hypervisor/cloud-hypervisor/issues/2525

**Comment by alyssais** [comment]
> @alyssais after some thinking and internal discussions, here is the
> design decision we took:
> https://github.com/cloud-hypervisor/cloud-hypervisor/issues/2525

Awesome, I think that's a good decision. :)
